sbndcode/Trigger/ArtdaqFragmentProducer_module.cc

```diff
@@ -0,0 +1,617 @@
+
+    num_pmt_wvf++;
+
+    double fStartTime = wvf.TimeStamp(); // in us
+    double fEndTime = double(wvf.size()) / fSampling + fStartTime; // in us
+
+    // create full waveform
+    std::vector<short> wvf_full; wvf_full.reserve((short)(3020*1e6/2));
+
+    if (fStartTime > fMinStartTime){
+      for (double i = fStartTime-fMinStartTime; i>0.; i-=(1./fSampling)){
+        wvf_full.push_back(fBaseline);
+      }
+    }
+
+    for(unsigned int i = 0; i < wvf.size(); i++) {
+      wvf_full.push_back(wvf[i]);
+    }
+
+    if (fEndTime < fMaxEndTime){
+      for (double i = fMaxEndTime-fEndTime; i>0.; i-=(1./fSampling)){
+        wvf_full.push_back(fBaseline);
+      }
+    }
+
+    // combine waveform with any other waveforms from same channel
+    int i_ch = -1.;
+    auto ich = std::find(channelList.begin(), channelList.end(), fChNumber);
+    if (ich != channelList.end()){
+      i_ch = ich - channelList.begin();
+    }
+    if (wvf_channel.at(i_ch).size() < wvf_full.size()){
+      if (fVerbose) std::cout<<"Full waveform -- Previous Channel" << fChNumber <<" Size: "<<wvf_channel.at(i_ch).size()<<"New Channel" << fChNumber <<" Size: "<<wvf_full.size()<<std::endl;
+      for(unsigned int i = wvf_channel.at(i_ch).size(); i < wvf_full.size(); i++) {
+        wvf_channel.at(i_ch).push_back(fBaseline);
+      }
+    }
+    for(unsigned int i = 0; i < wvf_full.size(); i++) {
+       wvf_channel.at(i_ch)[i] += (wvf_full[i] - fBaseline);
+    }
+
+    hist_id++;
+
+    wvfmHandle.clear();
+  } // waveform handle loop
+
+  // access hardware trigger information
+  std::vector<size_t> triggerIndex;
+  for(auto const& trigger : (*triggerHandle)) {
+    for (size_t idx = 0; idx < trigger.numPassed.size(); idx++) {
+      if (trigger.numPassed[idx] >= fMultiplicityThreshold) {
+        // save index
+        triggerIndex.push_back(idx);
+        // skip ahead by 5120 samples (downsampled by 4) before checking again
+        idx += 1280;
+      }
+    }
+  } // trigger handle loop
+
+  if (fVerbose) std::cout << "Number of PMT hardware triggers found: " << triggerIndex.size() << std::endl;
+
+  // set properties of fragment that are common to event
+  uint32_t nChannelsTotal = channelList.size();
+  //uint32_t nChannelsFrag = 15;
+  uint32_t nFrag = (uint32_t)nChannelsTotal/nChannelsFrag;
+  //uint32_t wfm_length = 5120; // ~10us, 2ns tick
+
+  // fragment properties
+  uint32_t sequenceIDVal = fEvent;
+
+  // create and populate common metadata
+  sbndaq::CAENV1730FragmentMetadata metadata;
+  metadata.nChannels = nChannelsFrag + 1; // 15 PMT channels + final channel to store beam window / trigger information
+  metadata.nSamples = wfm_length;
+
+  // fragment handle properties
+  uint32_t eventCounterVal = fEvent;
+  uint32_t boardIDVal = 0;
+  uint32_t triggerTimeTagVal = (uint32_t)CLHEP::RandFlat::shoot(&fTriggerTimeEngine, 0, 1e9);
+  uint32_t eventSizeVal = ((wfm_length * (nChannelsFrag+1)) / 2.) + sizeof(sbndaq::CAENV1730EventHeader);
```

Comment:
Aha, found the typo... We weren't dividing the event header size by 4.

```suggestion
  uint32_t eventSizeVal = ((wfm_length * (nChannelsFrag+1)) * sizeof(uint16_t) + sizeof(sbndaq::CAENV1730EventHeader)) / sizeof(uint32_t);
```

I don't know why but this field is stored in quad bytes. Not bytes. This is why we divide the length * channels chunk by 2 (each entry is 2 bytes) but we also need to do the same with the size of the event header, we want it in quad bytes, `sizeof` returns bytes.

I've changed the way we write it to try and make it more readable... Calculate the whole thing in bytes, then divide by the size of a quadbyte (4) but this is, in effect, the same as my suggestion the other day.

Reply:
@eyandel I will test this with the full system after lunch